NWChem ARMCI-MPI variant and optional TCE builds

jeffhammond opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
This is ready for review.

This adds two new features:
1. ARMCI-MPI can be used by Global Arrays, with `+armcimpi`.
2. Rarely-used TCE features are not compiled by default.  They can be added with `+extratce`.

The motivation for 1 is stability.  ARMCI-MPI performs better than MPI-TS (most of the time) and is easier to use than MPI-PR.  It also integrates nicely with MPI profiling and debugging tools.

The motivation for 2 is compilation time.  The extra TCE stuff more than doubles the compilation time.  The default here is consistent with manual builds of NWChem as well.

I'll note that `+extratce` includes a nontrivial portion of my PhD dissertation projects, so my choice to disable it by default is carefully considered.

I am working on GPU-related options now but that will go into a separate PR.

jeffhammond committed:
WIP NWChem with ARMCI-MPI and TCE optional

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
rename armci-mpi to armcimpi

jeffhammond committed:
add version for git

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
add ARMCI-MPI support to NWChem package

EXTERNAL_ARMCI_PATH needs to be set to the ARMCI-MPI package install prefix.
I could not get it from spec["armcimpi"] but it worked to use the dependent build environment method to export a generic environmental variable to use instead.

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
i suppose i can maintain NWChem as well

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
check rejects option that dozens of packages use

i do not have time to fight with this nonsense

Run . share/spack/setup-env.sh
==> Error: armcimpi version 'master' has extra arguments: 'branch'
Valid arguments for a url fetcher are:
    'url', 'sha256', 'md5', 'sha1', 'sha224', 'sha384', 'sha512', and 'checksum'
Error: Process completed with exit code 1.

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
style

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
ARMCI-MPI needs depends_on; the rest seems fixed

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

tldahlgren reviewed (dismissed):
LGTM

tldahlgren commented:
@yizeyi18 @jack-morrison @edoapra @zzzoom Would any of you be interested in reviewing this PR or at least the changes to `nwchem` since you made changes to the PR in the last year?

edoapra commented:
> @yizeyi18 @jack-morrison @edoapra @zzzoom Would any of you be interested in reviewing this PR or at least the changes to `nwchem` since you made changes to the PR in the last year?

No

zzzoom commented:
I can comment on the changes but I don't want to get tangled in whatever is happening here.

zzzoom commented on var/spack/repos/builtin/packages/armcimpi/package.py:
```diff
@@ -26,6 +26,8 @@ class ArmciMpi(AutotoolsPackage):
     variant("shared", default=True, description="Builds a shared version of the library")
     variant("progress", default=False, description="Enable asynchronous progress")
 
+    provides("armci")
```
You shouldn't need a virtual package for `armci` unless you're adding a second provider for it.

jeffhammond replied:
If Global Arrays (GA) becomes a separate Spack package, then it will be the second provider.  GA is currently not separate from NWChem but it can be and there are some apps that use it directly.  I intend to address this eventually but have higher priorities for now.

zzzoom replied:
Ok, fyi Global Arrays has a package, https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/globalarrays/package.py

jeffhammond replied:
Ah sorry. I guess I searched for "ga" and missed it.

jeffhammond replied:
a subsequent PR will reconcile the globalarrays and nwchem packages.  see https://github.com/spack/spack/commit/11a943c66846ed8c32103a17b22b2bcae8950d28 for details.

zzzoom commented on var/spack/repos/builtin/packages/nwchem/package.py:
```diff
@@ -35,6 +37,8 @@ class Nwchem(Package):
 
     variant("openmp", default=False, description="Enables OpenMP support")
     variant("mpipr", default=False, description="Enables ARMCI with progress rank")
+    variant("armcimpi", default=False, description="Enables ARMCI-MPI")
```
AFAICT from the documentation `ARMCI_NETWORK=MPI-PR` and `ARMCI_NETWORK=ARMCI` are mutually exclusive so `+mpipr` and `+armcimpi` should be a single multi-valued variant.

jeffhammond replied:
i don't know how to implement that.  i guess i'll figure it out.

zzzoom replied:
https://github.com/spack/spack/blob/49d3eb1723a2c7220883acfd2a24860a15ed4ef3/var/spack/repos/builtin/packages/globalarrays/package.py#L40

jeffhammond replied:
Thanks!

jeffhammond replied:
see 642958c6759e1cfe62866ed0e133fd045dae7df0

zzzoom commented on var/spack/repos/builtin/packages/nwchem/package.py:
```diff
@@ -35,6 +37,8 @@ class Nwchem(Package):
     variant("openmp", default=False, description="Enables OpenMP support")
     variant("mpipr", default=False, description="Enables ARMCI with progress rank")
+    variant("armcimpi", default=False, description="Enables ARMCI-MPI")
+    variant("extratce", default=False, description="Enables rarely-used TCE features")
```
I'd default to `True` so users don't lose any functionality when upgrading spack.

jeffhammond replied:
the impact on compilation time is profound.  `False` has been the default in NWChem for 20 years and no one has complained.  i dispute the backwards compatibility impact justifies the massive increase in compilation time.

yizeyi18 replied:
It may be better to give some explaination on what's disabled here, as we expect users of spack get all info they need by `spack info ...`, and NWChem's installation page what's non-default TCE feature is not mentioned.

jeffhammond replied:
https://nwchemgit.github.io/TCE.html#overview says:
> The distributed binary executables do not contain CCSDTQ and its derivative methods, owing to their large volume. The source code includes them, so a user can reinstate them by setenv CCSDTQ yes and recompile TCE module.

>The EA/IP-EOMCCSD methodologies are available in the 6.5 NWChem release. These implementation are available for the RHF type of the reference function. To enable the compilation of the EA/IP-EOMCCSD codes one has to set the following variable before the compilation of NWChem
> export EACCSD=y
> export IPCCSD=y

I don't document the `CCSDTLR` build option because I know what the user base is.  If somebody tries to run the method, the error message tells them it's not compiled and then they'll figure out what to do about it.

jeffhammond replied:
see 8acba5914fe9973d6574cc0a662b1ffbb8c460fa

jack-morrison commented:
> @yizeyi18 @jack-morrison @edoapra @zzzoom Would any of you be interested in reviewing this PR or at least the changes to `nwchem` since you made changes to the PR in the last year?

@zzzoom has already provided better comments than I'd be able to. Thanks for the ping but I don't think I can add value here.

jeffhammond committed:
fix ARMCI selection per review feedback from @zzzoom

https://github.com/spack/spack/pull/43883#discussion_r1585014147

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
address reviewer feedback from @yizeyi18

https://github.com/spack/spack/pull/43883#discussion_r1587228084

elaborate on what +extratce does, in terms of the NWChem build environment variables,
some of which are documented on https://nwchemgit.github.io/TCE.html.

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
style

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

tldahlgren reviewed (dismissed):
Let me know when this is ready to be merged.

zzzoom commented on var/spack/repos/builtin/packages/nwchem/package.py:
```diff
@@ -60,6 +72,7 @@ class Nwchem(Package):
     depends_on("blas")
     depends_on("lapack")
     depends_on("mpi")
+    depends_on("armcimpi", when="armci=armcimpi")
```
I tried to build all variants of `armci` and some deps are missing:
```suggestion
    depends_on("armcimpi", when="armci=armcimpi")
    depends_on("libfabric", when="armci=ofi")
    depends_on("rdma-core", when="armci=openib")
```
`armci=ofi` will still fail to build until https://github.com/GlobalArrays/ga/pull/324 gets patched in though.

jeffhammond committed:
Update var/spack/repos/builtin/packages/nwchem/package.py

Co-authored-by: Carlos Bederián <4043375+zzzoom@users.noreply.github.com>

zzzoom commented:
@tldahlgren LGTM
